### PR TITLE
BRS-502 Lock/Unlock records & fiscal years

### PR DESCRIPTION
**Pull request by cameronpettit**
This change includes the backend updates to enable locking/unlocking records.

There are 2 levels of locking: 1 on the record level against the record itself, and one at the fiscal year level, where all records within the fiscal year are locked against editing. If a record meets either one of these locking levels, it cannot be edited. 

New object `fiscalYearEnd` has been introduced with the following structure: 
```
pk: fiscalYearEnd
sk: 2023
isLocked: false
```
`sk` is the calendar year in which the fiscal year ends (2023 = 2022-04 to 2023-03). If `isLocked` is set to true, the fiscal year is locked against editing. 

New endpoints:
`GET ../fiscalYearEnd?fiscalYearEnd=2023` - get fiscalYearEnd object (to check locked status)
`POST ../fiscalYearEnd/lock?fiscalYearEnd=2023` - lock fiscalYearEnd object
`POST ../fiscalYearEnd/unlock?fiscalYearEnd=2023` - unlock fiscalYearEnd object

There are also new endpoints for `subarea`, which is the endpoint for creating/getting records.

`POST ../subarea/lock`
`POST ../subarea/unlock`

Both of the above requests must be accompanied by a body containing the following at a minimum:
```
subAreaID: 1234,
activity: Frontcountry Camping,
date: 202209
```
to uniquely identify the record to lock/unlock.

When a record is locked for the first time, the `isLocked: true` attribute is appended to the object. The absence of this attribute or setting `isLocked: false` indicates an unlocked record. 

When attempting to lock a record that does not yet exist, the API will create a blank record from the corresponding `config` object.

### CODE REVIEW

File: lambda/fiscalYearEnd/GET/index.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    };
+  }
+  return event.queryStringParameters.fiscalYearEnd;
+}
+
+async function getDateConfig(year) {
+  // check db for fiscalYearEnd object
+  try {
+    let fiscalYearEnd = await getOne('fiscalYearEnd', year);
+    // fiscalYearEnd = AWS.DynamoDB.Converter.unmarshall(fiscalYearEnd);
```
Comment (marklise) [inline lambda/fiscalYearEnd/GET/index.js:32]: prune.

File: lambda/subarea/POST/index.js [hunk 2]
```diff
@@ -20,40 +37,112 @@ exports.handler = async (event, context) => {
...
-      throw "Invalid request";
+    // check if fiscal year is locked
+    if (await checkFiscalYearLock(body)) {
+      return sendResponse(403, { msg: `This fiscal year has been locked against editing by the system administrator.` }, context)
+    }
+
+    // Disabling the ability to change config for now. 
+    // Request used to necessitate 'type = activity/config' as a queryParam (was future proofing).
+    // Queryparams no longer required. All info included in request body. 
+    // Refer to code prior to 2022-09-27 for handleConfig.
```
Comment (marklise) [inline lambda/subarea/POST/index.js:48]: 👍 Nice cleanup.

File: lambda/subarea/POST/index.js [hunk 3]
```diff
@@ -20,40 +37,112 @@ exports.handler = async (event, context) => {
...
+
+    await verifyBody(body);
+
+    // check if record is locked
+    const unlocking = lock === false ? true : false;
+    const existingRecord = await getOne(`${body.subAreaId}::${body.activity}`, body.date);
+    if (existingRecord?.isLocked && !unlocking) {
+      return sendResponse(403, { msg: 'Record is locked.' });
```
Comment (marklise) [inline lambda/subarea/POST/index.js:56]: HTTP 409 is a better response code.  You'll need to update your UT's accordingly.

File: lambda/subarea/POST/index.js [hunk 4]
```diff
@@ -20,40 +37,112 @@ exports.handler = async (event, context) => {
...
+    Key: {
+      pk: { S: record.pk },
+      sk: { S: record.sk }
+    },
+    UpdateExpression: 'set isLocked = :isLocked',
+    ExpressionAttributeValues: {
+      ':isLocked': { BOOL: lock }
+    },
+    ReturnValues: 'ALL_NEW'
+  };
```
Comment (marklise) [inline lambda/subarea/POST/index.js:112]: We may want to put a `ConditionExpression` of `isLocked = !lock`.  That way when you fail you can return `HTTP 409`.

File: lambda/fiscalYearEnd/GET/index.js (hunk 1 above)
Comment (cameronpettit) [reply lambda/fiscalYearEnd/GET/index.js]: Fixed in `89812d6`.

File: lambda/subarea/POST/index.js (hunk 2 above)
Comment (cameronpettit) [reply lambda/subarea/POST/index.js:48]: Fixed in `89812d6`.

File: lambda/subarea/POST/index.js (hunk 3 above)
Comment (cameronpettit) [reply lambda/subarea/POST/index.js]: Fixed in `89812d6`.

File: lambda/subarea/POST/index.js (hunk 4 above)
Comment (cameronpettit) [reply lambda/subarea/POST/index.js]: Fixed in `89812d6`.

**Review by marklise — APPROVED** [review]

**Review by danieltruong — APPROVED** [review]

**Pull request merged**